src/com/rabbitmq/client/impl/recovery/AutorecoveringConnection.java

```diff
@@ -428,18 +470,43 @@ private void recoverBlockedListeners() {
         }
     }
 
-    private void recoverConnection() throws IOException, InterruptedException {
-        boolean recovering = true;
-        while (recovering) {
+	//Returns true if the connection was recovered, 
+	//false if application initiated shutdown while attempting recovery.  
+    private boolean recoverConnection() throws IOException, InterruptedException {
```

Comment:
I don't see another way to do this without restructuring beginAutomaticRecovery.  I cannot check manuallyClosed after calling recoverConnection() without a potential race condition in which I leak the connection object.  beginAutomaticRecovery needs to know if a new connection was created.